src/levelset/levelSetSegmentationObject.cpp

```diff
@@ -298,59 +298,291 @@ std::array<double, 3> LevelSetSegmentationSurfaceInfo::evalNormal(const std::arr
+    m_surface->getVertexCoords(segmentVertexIds.size(), segmentVertexIds.data(), segmentVertexCoors);
+
+    BITPIT_CREATE_WORKSPACE(lambda, double, nSegmentVertices, ReferenceElementInfo::MAX_ELEM_VERTICES);
+    (*projectionPoint) = CGElem::projectPointPolygon(point, nSegmentVertices, segmentVertexCoors, lambda);
+
+    (*projectionNormal) = lambda[0] * computeSegmentVertexNormal(segmentItr, 0, true);
+    for (std::size_t i = 1; i < nSegmentVertices; ++i) {
+        (*projectionNormal) += lambda[i] * computeSegmentVertexNormal(segmentItr, i, true);
+    }
+    (*projectionNormal) /= norm2(*projectionNormal);
+}
+
 /*!
  * Evaluate the projection of the given point on the specified segment.
  *
  * @param[in] point are the coordinates of point
  * @param[in] segmentItr is an iterator pointing to the closest segment
+ * @param[out] projectionPoint The coordinates of the projection point on the surface.
+ * @param[out] projectionNormal The coordinates of the norrmal to the surface vector on the surface
+ */
+void LevelSetSegmentationSurfaceInfo::evalLowOrderProjection(const std::array<double, 3> &point,
+                                                             const SegmentConstIterator &segmentItr,
+                                                             std::array<double, 3> *projectionPoint,
+                                                             std::array<double, 3> *projectionNormal) const
+{
+    const Cell &segment = *segmentItr;
+    ElementType segmentType = segment.getType();
+    switch (segmentType) {
+
+    case ElementType::VERTEX:
+    {
+        evalProjectionOnVertex(point, segmentItr, projectionPoint, projectionNormal);
+        return;
+    }
+
+    case ElementType::LINE:
+    {
+        evalLowOrderProjectionOnLine(point, segmentItr, projectionPoint, projectionNormal);
+        return;
+    }
+
+    case ElementType::TRIANGLE:
+    {
+        evalLowOrderProjectionOnTriangle(point, segmentItr, projectionPoint, projectionNormal);
+        return;
+    }
+
+    default:
+    {
+        evalLowOrderProjectionOnPolygon(point, segmentItr, projectionPoint, projectionNormal);
+        return;
+    }
+
+    }
+}
+
+/*!
+ * Evaluate the baricentric coordinates of the given point on the specified segment.
+ *
+ * @param[in] point are the coordinates of point
+ * @param[in] segmentItr is an iterator pointing to the closest segment
  * @param[out] lambda on output will contain the barycentric coordinates of the projection point
- * @return The coordinates of the projection point.
  */
-std::array<double, 3> LevelSetSegmentationSurfaceInfo::evalProjection(const std::array<double, 3> &point,
-                                                                      const SegmentConstIterator &segmentItr,
-                                                                      double *lambda) const
+void LevelSetSegmentationSurfaceInfo::evalBarycentricCoordinates(const std::array<double, 3> &point,
```

Comment:
Can we move evalBarycentricCoordinates in SurfaceKernel (and isnce we are at it, also in LineKernel)?

Reply:
Done